hockeypuck_rock/delete_keys.py

```
@@ -0,0 +1,230 @@
+            WHERE keys.rfingerprint = REVERSE(deleted_keys.fingerprint);
+        """
+        )
+
+        cursor.connection.commit()
+
+        logging.info("Deletion process completed successfully.")
+    except psycopg2.Error as e:
+        raise DatabaseOperationError(f"Error executing SQL commands: {e}") from e
+
+
+def main() -> None:
+    """Main entrypoint.
+
+    Raises:
+        InvalidFingerprintError: if any of the fingerprints are invalid.
+        KeyDeletionError: if any of the operations fail.
+    """
+    parser = argparse.ArgumentParser(
+        description="Delete keys from the Hockeypuck Postgres database by fingerprint."
+    )
+    parser.add_argument(
+        "--fingerprints", required=True, help="Comma-separated list of fingerprints to delete"
+    )
+    parser.add_argument("--comment", required=True, help="Comment associated with the deletion")
+    args = parser.parse_args()
+    fingerprints = args.fingerprints.split(",")
+    invalid_fingerprints = []
+    for fingerprint in fingerprints:
+        # fingperints are usually of length 40 or 64 depending on the hash algorithm, and
+        # consist of hexadecimal characters only.
+        if not re.fullmatch(r"[0-9A-Fa-f]{40}|[0-9A-Fa-f]{64}", fingerprint):
+            logging.error("Invalid fingerprint format: %s", fingerprint)
+            invalid_fingerprints.append(fingerprint)
+    if invalid_fingerprints:
+        raise InvalidFingerprintError(invalid_fingerprints)
+
+    comment = args.comment
+
+    try:
+        with get_db_connection() as conn:
+            with conn.cursor() as cursor:
+                delete_fingerprints(cursor, fingerprints, comment)
+                remove_ptree_data()
+                invoke_rebuild_prefix_tree()
+    except (
+        DatabaseConnectionError,
+        DatabaseOperationError,
+        PtreeDataError,
+        PrefixTreeRebuildError,
+    ) as e:
+        logging.error("Unable to delete keys: %s", e)
```

And you can re-raise the exception here

---

already did